crates/worker/src/partition_processor_manager.rs

```diff
@@ -645,35 +744,48 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
             CreateSnapshot(partition_id, sender) => {
                 self.running_partition_processors
                     .get(&partition_id)
-                    .map(|store| store.handle.create_snapshot(Some(sender)));
+                    .map(|store| {
+                        let ProcessorStatus::Started(state) = store else {
+                            return None;
+                        };
+
+                        Some(state.handle.create_snapshot(Some(sender)))
+                    });
             }
             GetState(sender) => {
                 let _ = sender.send(self.get_state());
             }
         }
     }
 
-    async fn on_control_processors(
+    #[instrument(parent=None, skip_all)]
+    fn on_control_processors(
         &mut self,
-        control_processor: Incoming<ControlProcessors>,
-    ) -> Result<(), Error> {
-        let control_processors = control_processor.into_body();
-
-        self.metadata
-            .wait_for_version(
-                MetadataKind::Logs,
-                control_processors.min_logs_table_version,
-            )
-            .await?;
-        let partition_table = self
-            .metadata
-            .wait_for_partition_table(control_processors.min_partition_table_version)
-            .await?
-            .into_arc();
-
-        for control_processor in control_processors.commands {
-            self.on_control_processor(control_processor, &partition_table)
-                .await?;
+        control_processors: ControlProcessors,
+        effects: mpsc::UnboundedSender<ManagerEffect>,
+    ) -> Result<(), ShutdownError> {
+        self.control_processors = None;
+        // version already satisfied so we can directly apply th effects
+        if control_processors.min_logs_table_version <= self.metadata.logs_version()
+            && control_processors.min_partition_table_version
+                <= self.metadata.partition_table_version()
+        {
+            let partition_table = self.metadata.partition_table_snapshot();
+
+            for control_processor in control_processors.commands {
+                effects
+                    .send(ManagerEffect {
+                        parent: Span::current(),
+                        effect: ProcessorEffect::ControlProcessor {
+                            control_processor,
+                            partition_table: Arc::clone(&partition_table),
+                        },
+                    })
+                    .map_err(|_| ShutdownError)?
+            }
```

Comment:
Could we call `on_control_processor` directly here w/o going through the `effects` channel? W/ your changes, I assume that `on_control_processor` will be really fast to complete.

Reply:
Yeah, it's definitely possible to avoid the effects channel all together and make the call directly. I will change it